File: Python/programmers/kakao/level2/오픈채팅방.py
from collections import defaultdict

def solution(record):
    tmp = []
    name = defaultdict(str)
    for i in range(len(record)):
        cmd = record[i].split()
        if len(cmd) == 3:
            state, ids, nick = cmd[0], cmd[1], cmd[2]
            name[ids] = nick
            if state == "Enter":
                tmp.append([state, ids])
        else:
            state, ids = cmd[0], cmd[1]
            tmp.append([state, ids])
    
    answer = []
    for t in tmp:
        if t[0] == "Enter":
            s = f"{name[t[1]]}님이 들어왔습니다."
        else:
            s = f"{name[t[1]]}님이 나갔습니다."
        answer.append(s)
    
    return answer


record = ["Enter uid1234 Muzi", "Enter uid4567 Prodo","Leave uid1234","Enter uid1234 Prodo","Change uid4567 Ryan"]
print(solution(record)) # ["Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다."]

File: Java/programmers/kakao/level2/오픈채팅방.java
package CodeTest.Java.programmers.kakao.level2;

import java.util.*;

public class 오픈채팅방 {
    public String[] solution(String[] record) {
        List<String[]> tmp = new ArrayList<>();
        Map<String, String> map = new HashMap<>();

        for (int i=0; i<record.length; i++) {
            String[] cmd = record[i].split(" ");
            if (cmd.length == 3) {
                String state = cmd[0], ids = cmd[1], nick = cmd[2];
                map.put(ids, nick);
                if (state.equals("Enter")) {
                    tmp.add(new String[]{state, ids});
                }
            } else {
                String state = cmd[0], ids = cmd[1];
                tmp.add(new String[]{state, ids});
            }
        }

        String[] answer = new String[tmp.size()];
        
        for (int i=0; i<tmp.size(); i++) {
            if (tmp.get(i)[0].equals("Enter")) {
                answer[i] = map.get(tmp.get(i)[1]) + "님이 들어왔습니다.";
            } else {
                answer[i] = map.get(tmp.get(i)[1]) + "님이 나갔습니다.";
            }
        }

        return answer;
    }
}
